core/sail/api/src/main/java/org/eclipse/rdf4j/common/concurrent/locks/AbstractReadWriteLockManager.java

```diff
@@ -86,26 +98,121 @@ protected void waitForActiveWriter() throws InterruptedException {
 	 * @throws InterruptedException
 	 */
 	protected void waitForActiveReaders() throws InterruptedException {
-		activeReaders.waitForActiveLocks();
+		while (isReaderActive()) {
+			Thread.onSpinWait();
+		}
 	}
 
 	/**
 	 * Creates a new Lock for reading and increments counter for active readers. The lock is tracked if lock tracking is
 	 * enabled. This method is not thread safe itself, the calling method is expected to handle synchronization issues.
```

Comment:
Is this still true given that you're now relying on coordination between two LongAdders for synchronization?

Reply:
Haven't updated all the documentation yet.

Reply:
Also I haven't incorporated any of the lock tracking back into the read-write lock code. 

Got a bit stuck right now because I discovered I need to write an entire suite of tests for verifying what features the previous code supported when it comes to tracking locks, automatically releasing GCed locks and detecting deadlocks.